framework/src/node/consensus/consensus.ts

```diff
@@ -641,17 +648,58 @@ export class Consensus {
 		}
 	}
 
+	private _verifyStateRoot(block: Block, stateRoot: Buffer): void {
+		if (!block.header.stateRoot || !stateRoot.equals(block.header.stateRoot)) {
+			throw new Error(
+				`State root is not valid for the block with id: ${block.header.id.toString('hex')}`,
+			);
+		}
+	}
+
 	private async _deleteBlock(block: Block, saveTempBlock = false): Promise<void> {
 		if (block.header.height <= this._chain.finalizedHeight) {
 			throw new Error('Can not delete block below or same as finalized height');
 		}
 
 		// Offset must be set to 1, because lastBlock is still this deleting block
 		const stateStore = new StateStore(this._db);
-		await this._chain.removeBlock(block, stateStore, { saveTempBlock });
+		const currentState = await this._finalizeStore(stateStore);
+		await this._chain.removeBlock(block, currentState, { saveTempBlock });
 		this.events.emit(CONSENSUS_EVENT_BLOCK_DELETE, block);
 	}
 
+	private async _finalizeStore(stateStore: StateStore, finalize = true): Promise<CurrentState> {
```

Comment:
do we need `, finalize = true`? name is finalize and having flag seems bit strange.

Reply:
maybe `_prepareStore`, and then we can pass finalize flag ?